addons/pos_adyen/models/pos_payment_method.py

```diff
@@ -0,0 +1,94 @@
+# coding: utf-8
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import json
+import logging
+import pprint
+import random
+import requests
+import string
+
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError
+
+_logger = logging.getLogger(__name__)
+
+class PosPaymentMethod(models.Model):
+    _inherit = 'pos.payment.method'
+
+    use_payment_terminal = fields.Selection(selection_add=[('adyen', 'Adyen')])
+    adyen_api_key = fields.Char(string="Adyen API key", help='Used when connecting to Adyen: https://docs.adyen.com/user-management/how-to-get-the-api-key/#description', copy=False)
+    adyen_terminal_identifier = fields.Char(help='[Terminal model]-[Serial number], for example: P400Plus-123456789', copy=False)
+    adyen_test_mode = fields.Boolean(help='Run transactions in the test environment.')
+    adyen_latest_response = fields.Char(help='Technical field used to buffer the latest asynchronous notification from Adyen.', copy=False, groups='base.group_erp_manager')
+    adyen_latest_diagnosis = fields.Char(help='Technical field used to determine if the terminal is still connected.', copy=False, groups='base.group_erp_manager')
+
+    def _is_write_forbidden(self, fields):
+        whitelisted_fields = set(('adyen_latest_response', 'adyen_latest_diagnosis'))
+        return super(PosPaymentMethod, self)._is_write_forbidden(fields - whitelisted_fields)
+
+    def _adyen_diagnosis_request_data(self, pos_config_name):
+        service_id = ''.join(random.choices(string.ascii_letters + string.digits, k=10))
+        return {
+            "SaleToPOIRequest": {
+                "MessageHeader": {
+                    "ProtocolVersion": "3.0",
+                    "MessageClass": "Service",
+                    "MessageCategory": "Diagnosis",
+                    "MessageType": "Request",
+                    "ServiceID": service_id,
+                    "SaleID": pos_config_name,
+                    "POIID": self.adyen_terminal_identifier
+                },
+                "DiagnosisRequest": {
+                    "HostDiagnosisFlag": False
+                }
+            }
+        }
+
+    def get_latest_adyen_status(self, pos_config_name):
+        self.ensure_one()
+        latest_response = self.sudo().adyen_latest_response
+        latest_response = json.loads(latest_response) if latest_response else False
+        self.sudo().adyen_latest_response = ''  # avoid handling old responses multiple times
```

Comment:
in adyen_pay_data, we are sending a 'SaleTransactionID' which correspond to the order.uid, shouldn't we check that the response is for this 'SaleTransactionID' ? (if it is possible)

Reply:
Indeed, but this check happens in the frontend (in `_poll_for_response`). We check using the `ServiceID` instead. It's a random number that's generated with each request done to Adyen. Adyen will copy this number in the response. Since there can only be one order active at a time per terminal that's enough I think.